docs/rfcs/xxxx-integrations-in-operator.md

```diff
@@ -0,0 +1,265 @@
+experimental feature is available.
+
+### Integration validation
+
+The initial implementation of integrations support will have no knowledge of
+what integrations exist. As a result, the `spec.type` and `spec.config` fields
+for an Integration MUST be configured correctly for an integration to work.
+Users must refer to documentation to discover how `type` should be configured
+for their specific integration, and what settings are valid for the `config`
+block. Configuration errors will only surface as runtime errors from the
+deployed agent.
+
+Future versions of the Operator may:
+
+* Add knowledge for some integrations and validate `type` and `config`
+  accordingly (though breaking changes to the config at the Agent level may
+  introduce extra complexity to this).
+
+* Update the `status` field of the root GrafanaAgent resource during reconcile
+  to expose any reconcile or runtime errors.
+
+### Additional settings for the Integration CRD
+
+Some integrations may require changes to the deployed Pods to function
+properly. Integrations will additionally support declaring `volumes`,
+`volumeMounts`, `secrets` and `configMaps`. These fields will be merged with
+the fields of the same name from the root GrafanaAgent resource when creating
+integration pods:
+
+> ```yaml
+> apiVersion: monitoring.grafana.com/v1alpha1
+> kind: Integration
+> metadata:
+>   name: kafka
+>   namespace: default
+> spec:
+>   name: kafka_exporter
+>   type:
+>     hasMetrics: true
+>   config: |
+>     ca_file: /etc/grafana-agent/secrets/kafka-ca-file
+>     # ...
+>   # Same "secrets" field present in GrafanaAgent.spec, where each secret
+>   # is loaded from the same namespace and gets exposed at
+>   # /etc/grafana-agent/secrets/<secret name>
+>   secrets: [kafka-ca-file]
+> ```
+
+### Sending telemetry from integrations
+
+Because the operator will not have any knowledge about individual integrations, it
+also doesn't know how integrations generate telemetry data. Users must manally
+configure an integration to send its data to the appropriate instance.
+
+Users can refer to MetricsInstances and LogsInstance from the same resource
+hierarchy by `<namespace>/<name>` in their integration configs. This
+configuring `autoscrape` for collecting metrics from an exporter-based
+integration:
+
+> ```yaml
+> apiVersion: monitoring.grafana.com/v1alpha1
+> kind: MetricsInstance
+> metadata:
+>   name: primary
+>   namespace: default
+> spec:
+>   remoteWrite:
+>   - url: http://prometheus:9090/api/v1/write
+> ---
+> apiVersion: monitoring.grafana.com/v1alpha1
+> kind: Integration
+> metadata:
+>   name: mysql
+>   namespace: default
+> spec:
+>   name: mysqld_exporter
+>   config:
+>     autoscrape:
+>       enable: true
+>       # MetricsInstance <namespace>/<name> to send metrics to
+>       metrics_instance: default/primary
+>     data_source_name: root@(mysql.default:3306)/
+>     disable_collectors: [slave_status]
+> ```
+
```

Comment:
For something like the `eventhandler` integration how would this work?

Would the config look something like this?

```yaml
apiVersion: monitoring.grafana.com/v1alpha1
kind: Integration
metadata:
   name: eventhandler
   namespace: default
 spec:
   name: eventhandler
   config:
       logs_instance: ?
       misc_eventhandler_config ...
```

I guess some sort of local "ghost' logs instance would be enabled (with what config?) and the integration could use that one?

Reply:
You'd set `logs_instance` to the `<namespace>/<name>` of the LogsInstance CR in your resource hierarchy. Does that make sense? 

```yaml 
apiVersion: monitoring.grafana.com/v1alpha1
kind: GrafanaAgent
metadata:
  name: grafana-agent-example
  namespace: default
  labels:
    app: grafana-agent-example
spec:
  logs:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-example
  integrations:
    instanceSelector:
      matchLabels:
        agent: grafana-agent-example
---
apiVersion: monitoring.grafana.com/v1alpha1
kind: LogsInstance
metadata:
  name: primary
  namespace: default
  labels:
    agent: grafana-agent-example
spec:
  clients:
  - url: http://loki:8080/loki/api/v1/push
---
apiVersion: monitoring.grafana.com/v1alpha1
kind: Integration
metadata:
   name: eventhandler
   namespace: default
spec:
   name: eventhandler
   config:
     logs_instance: default/primary
     misc_eventhandler_config: ...
```

Reply:
Sorry if this was already assumed, but just in case: eventhandler works by handing off log entries to a channel managed by a logs instance running in the _same_ Agent (in other words, the same Agent must run a logs instance and eventhandler integration).

The integration can be modified to not do this, and maybe write out to a file or something...

But given the above IIUC: 

- User defines a `LogsInstance`
- Operator reconciles and creates an Agent DaemonSet
- User creates a "Deployment" integration (so `allNodes: false`)
- Operator reconciles and creates an Agent Deployment running the integration
- Operator also detects LogsInstance and so will create config enabling a `logs` instance on that integrations Agent with the same config as the LogsInstance CR? 

What happens if several `LogsInstance`s are defined? Are these all appended to the integrations Agent's config? (sorry if this doesn't make sense...)

Reply:
I'll try to make this more obvious in the doc, but the idea is that the above config generates the following agent config file:

```yaml
logs:
  configs:
  - name: default/primary
    clients:
    - url: http://loki:8080/loki/api/v1/push

integrations:
  eventhandler:
    logs_instance: default/primary 
    misc_eventhandelr_config: ... 
```

Any LogsInstance or MetricsInstance in the same hierarchy as the integration will be injected into the generated config file so integrations can use them.

Reply:
Gotcha understood